### PR TITLE
[Composable Kernel] Integrate xdlops fwd kernel

### PR DESCRIPTION
**Pull request by ltqin**
This PR integrate composable kernel fwd v4r4r4  into MIOpen
test command:
./build/bin/MIOpenDriver convfp16 -n 128 -c 128 -H 14 -W 14 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1   -t 1 -F 1 -O NHWC -I NHWC -f NHWC -S 121

### CODE REVIEW

**Comment by junliume** [comment]
@ltqin could you please add the test to test\CMakeLists.txt?
CC: @atamazov

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (junliume) [inline src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:138]:
```suggestion
    bool AThreadTransferSrcResetCoordinateAfterRun                = true;
```
tidy complaint

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (junliume) [inline src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:147]:
```suggestion
    bool BThreadTransferSrcResetCoordinateAfterRun                = true;
```
tidy complaint

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (junliume) [inline src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:444]: implicit conversion bool -> 'int' [readability-implicit-bool-conversion,-warnings-as-errors]
HasMainKBlockLoop},
 ^
static_cast<int>()

**Comment by atamazov** [comment]
- motivation/goals?
- urgency/target release?
- side effects?

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp]: done

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp]: done

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:444]: done

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: #1270 spec opened

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [inline src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:324]: Please move definitions of all non-small methods (e.g. > 5 lines) to the corresponding .cpp file (filename should match header name).

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:1581]: The new solver is applicable for gfx90a. Add `GFX90A_ENABLED`

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:138]: :red_circle: The new solver is applicable for gfx90a and FP16. Please take care about this. Either provide FP16 ALT implementations in kernels, or disable when necessary.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/CMakeLists.txt
Comment (ltqin) [reply test/CMakeLists.txt]: done

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (ltqin) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp]: done

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: yes, KernelBuildParameters is good, but it can not be used in CK

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:324]: If there is no virtual function and no information hiding, is it necessary to separate?

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: Why?

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:324]: There are so many implementation details that they clutter up the interface and hinder readability. Perhaps this is a matter of taste, let's ask others.

CC @JehandadKhan @junliume

Comment (JehandadKhan) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:324]: I would agree with @atamazov  on the grounds that moving definitions to cpp files improves compilation and is better design.

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (atamazov) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:138]: Pls let me know if you need more assistance on this.

Comment (ltqin) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:141]: pls help me by checking commit "e30aa9e"

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: This file is in the path of src/composable_kernel that can synchronize with compoable_kernel repository. but "kernel_build_params.hpp"  is in the path src/include/miopen that We don't want to track  in CK

Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: Then something is wrong with integration of composable_kernel repository. I think we shouldn't merge this as this adds postponed work to the existing massive.

CC @JehandadKhan @pfultz2 @junliume

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (atamazov) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:141]: Looks good!

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp]: done

**Comment by atamazov** [comment]
@ltqin Force-pushing is NOT recommended due to several reasons. It can be more or less safely used only if the branch has been published for review yet.

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: not resolved

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:324]: not resolved

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: done

**Comment by ltqin** [comment]
> @ltqin Force-pushing is NOT recommended due to several reasons. It can be more or less safely used only if the branch has been published for review yet.

Got it, I should use "revert"

**Comment by atamazov** [comment]
> Got it, I should use "revert"

No, just make necessary changes and push to remote. If someone else is working on your branch and he/she added some commits, then git will refuse. In this case you need to _pull from remote and resolve merge conflicts_ (if any), and then repeat push.

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (atamazov) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: @ltqin Good, thanks! I will get back to reviewing of this as soon as I finish with 5.0 and other urgent stuff.

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:140]: done

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (atamazov) [inline src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:1]: Please add our standard copyright header comment.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/CMakeLists.txt
Comment (atamazov) [inline test/CMakeLists.txt:1582]:
```
set(CONV_FORWARD_ONLY_NHWC
    --disable-backward-data --disable-backward-weights
    --in_layout NHWC --fil_layout NHWC --out_layout NHWC)
```
and use it in the new test cases instead of "--disable-backward-data --disable-backward-weights --in_layout NHWC --fil_layout NHWC --out_layout NHWC"

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by atamazov** [comment]
[This info](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1260#issuecomment-963484199) is still missing in the PR.

File: src/composable_kernel/host/solver/include/convolution_problem_descriptor.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/convolution_problem_descriptor.hpp:3]: ConvolutionProblemDescriptor is not generic problem(miopen), it is defined and used in CK

File: src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp
Comment (ltqin) [reply src/composable_kernel/host/solver/include/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.hpp:10]: done,  data_type_enum.hpp is now included in convolution_problem_descriptor.hpp, can remove it

File: test/CMakeLists.txt
Comment (ltqin) [reply test/CMakeLists.txt]: done

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (ltqin) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:26]: done

**Comment by ltqin** [comment]
> [This info](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1260#issuecomment-963484199) is still missing in the PR.

The goals is test out the way to add xdlops kernel of CK into miopen, CK can quickly  add kernels to new device and new feature.  so this PR is not urgency. 
I have no idea of side effects about this pr.

**Comment by zjing14** [comment]
@atamazov 
- motivation/goals
This PR is to integrate first CK:iGemm for MI100/200 as a demo to evaluate our kernel and compatibility of ck compilation-flags/tuning mechamism to the MIOpen solver for future integration works
- urgency/target release
Not urgent and target release
- how tested?
added tests in CMakeLists
- side effects
No side effect. Asm-iGemm is still winner.

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:175]:
```suggestion
        // This kernel uses int32_t for memory offset, which covers 2GB of memory maximum.
```

**Review by atamazov — CHANGES_REQUESTED** [review]
(Partial)

**Comment by atamazov** [comment]
@zjing14 @ltqin Please correct me if I am wrong:

> * motivation/goals
>   * This PR is to integrate first CK:iGemm for MI100/200 as a demo to evaluate our kernel and compatibility of ck compilation-flags/tuning mechamism to the MIOpen solver for future integration works.

If this is the case, then it is worth planning either further development (from demo to production) or retirement of the new Solver (before merging it). Thoughts?

> * how tested?
>   * added tests in CMakeLists

This is good, but does not seem enough to ensure the production quality. I guess that actually the new kernels were thoroughly tested during development, right?

>* side effects
>   * No side effect. Asm-iGemm is still winner.

This adds a new tunable Solver. Therefore it affects tuning time, because Tuna runs all solvers (as if `MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS` is set). @JehandadKhan 

As we do not expect performance progressions from this Solver, I highly recommend disabling it by default.

File: src/composable_kernel/composable_kernel/include/tensor_operation/blockwise_gemm_dlops_v3.hpp
Comment (atamazov) [inline src/composable_kernel/composable_kernel/include/tensor_operation/blockwise_gemm_dlops_v3.hpp:114]: It seems like this PR contains a lot of useful CK-wide changes. Maybe it is worth moving them to a separate PR?

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (ltqin) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:290]: There is a size check in the IsApplicable function

Comment (atamazov) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:290]: Can you please provide a link? I um unable to find it.

File: src/composable_kernel/composable_kernel/include/problem_transform/transform_backward_weight_convolution_into_gemm_v4r4r2_atomic_nchw_kcyx_nkhw.hpp
Comment (ltqin) [reply src/composable_kernel/composable_kernel/include/problem_transform/transform_backward_weight_convolution_into_gemm_v4r4r2_atomic_nchw_kcyx_nkhw.hpp:1]: done

File: src/composable_kernel/host/driver_offline/include/debug.hpp
Comment (ltqin) [reply src/composable_kernel/host/driver_offline/include/debug.hpp]: done

File: test/CMakeLists.txt
Comment (ltqin) [reply test/CMakeLists.txt:1591]: have merged latest develop

File: src/solver.cpp
Comment (ltqin) [reply src/solver.cpp:496]: done

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (ltqin) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp:36]: I don't know why put  "specific" headers at the first. It seems opposite to habit

File: src/include/miopen/sqlite_db.hpp
Comment (ltqin) [reply src/include/miopen/sqlite_db.hpp:264]: Should I revert sqlite_db.hpp?  actually, clang-format-10 didn't deal with sqlite_db.hpp before.

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (ltqin) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:290]: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/ck_fwd_xdlops/src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp#L178

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (ltqin) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:282]: Mainly for readability

File: src/include/miopen/sqlite_db.hpp
Comment (atamazov) [reply src/include/miopen/sqlite_db.hpp:264]: @ltqin Yes, please. I see similar strange behavior of clang-format-10 on may local machine (it formats files that I didn't touched).

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (atamazov) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:290]: The connection is non-trivial for the reader. Please add comment saying that IsApplicable checks that tensor size fits into int (N*H*W*C*element_size <= INT_MAX) and therefore N*H*W also fits int here.

File: src/include/miopen/sqlite_db.hpp
Comment (atamazov) [reply src/include/miopen/sqlite_db.hpp:264]: @JehandadKhan @pfultz2 gentle ping wrt https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1260#discussion_r756423900

File: src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp
Comment (ltqin) [reply src/solver/conv_ck_igemm_fwd_v4r4r4_xdlops_nhwc.cpp]: done

File: src/include/miopen/sqlite_db.hpp
Comment (ltqin) [reply src/include/miopen/sqlite_db.hpp]: done

Comment (JehandadKhan) [reply src/include/miopen/sqlite_db.hpp]: I think if there is a formatting issue (ie file skipped in formatting tests) then we should address it in a separate PR. So @ltqin can NOT add the missing file in this PR and maintain the state as is and then a separate PR can fix that issue in an isolated manner.

File: src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp
Comment (ltqin) [reply src/composable_kernel/host/solver/src/conv_igemm_fwd_v4r4r4_xdlops_nhwc_kyxc_nhwk.cpp:290]: done

**Comment by atamazov** [comment]
If we do not want this to be included into 5.0, then let's put this on hold until release branch is cut.

**Pull request closed without merging**
